src/gurt/fault_inject.c

```diff
@@ -99,31 +149,28 @@ fault_attr_set(uint32_t fault_id, struct d_fault_attr_t fa_in, bool take_lock)
 	if (take_lock)
 		D_RWLOCK_WRLOCK(&d_fi_gdata.dfg_rwlock);
 
-	if (fault_id >= d_fi_gdata.dfg_fa_capacity) {
-		new_capacity = fault_id + 1;
-		D_REALLOC_ARRAY(new_fa_arr, d_fi_gdata.dfg_fa, new_capacity);
-		if (new_fa_arr == NULL)
-			D_GOTO(out_unlock, rc = -DER_NOMEM);
-
-		start = new_fa_arr + d_fi_gdata.dfg_fa_capacity;
-		num_bytes = sizeof(*new_fa_arr)
-			    * (new_capacity - d_fi_gdata.dfg_fa_capacity);
-		memset(start, 0, num_bytes);
-		d_fi_gdata.dfg_fa = new_fa_arr;
-		d_fi_gdata.dfg_fa_capacity = new_capacity;
-	}
-
-	fault_attr = d_fi_gdata.dfg_fa[fault_id];
-	if (fault_attr == NULL) {
-		fault_attr = new_fault_attr;
+	rlink = d_hash_rec_find(&d_fi_gdata.dfg_fa_table, &fault_id,
+				sizeof(fault_id));
+	if (rlink == NULL) {
+		fault_attr = &new_rec->fa_attr;
```

Comment:
why not D_ALLOC_PTR(new_rec) here instead? can then just rely on rc != 0 for freeing and no need for 'should_free' var

Reply:
D_ALLOC_PTR() needs to be outside the locking call on line 150: D_RWLOCK_WRLOCK(&d_fi_gdata.dfg_rwlock);

otherwise there's a deadlock because D_ALLOC_PTR() also takes that lock.